semgrep-core/Analyzing/Constant_propagation.ml

```diff
@@ -304,22 +313,21 @@ let propagate_basic lang prog =
          | N (Id (id, id_info)) ->
              (match find_id env id id_info with
               | Some literal ->
-                  id_info.id_constness := Some (Lit literal)
+                  join_constness_ref id_info.id_constness (Lit literal)
```

Comment:
But here id_constness should always be None no? this join is really a set_constness_ref?
Or is there a situation where it can be already set to a Some?

Reply:
Note that in the first step, `var_stats`, I go through all assignments `x = E` and assign `x` the constness `NotCst`. Here we are visiting a name, but we don't know whether this name is acting as a label (l-value) or as an r-value. If it were an r-value then it's correct that constness would be `None`, but it is an l-value then it would be `NotCst`. Perhaps I should document this better!

Reply:
Hmmm, maybe this is a bit hacky, it may be more clear if I change the way the AST is visited.

Reply:
The way I do it with `join_constness_ref` is a bit tricky but right now I think it may be the simplest way. The visitor for the AST doesn't allow to visit r-value expressions only I think. It would be easier if we update constness using IL, since in IL it's very clear where a variable is being read (that's where we want to update the constness) and where it acts as a label (here we don't want to set the constness). But right now the AST-to-IL transformation only works for individual functions and not for entire programs, so we would need to extend it first.

What do you think?

Reply:
yep, would be better to do that in IL where lvalue/rvalue is better separated.
Also right now we don't really detect well lvalue in the AST because you can do tuple assignement like (a, b, c) = (1, 2, 3).
In the IL it would be simpler because we could decompose those as a = 1; b = 2; c = 3, or if the rhs is not a tuple as in
(a, b, c) = foo(), we should generate in the IL   _tmp = foo(); a = _tuple_access(_tmp, 1); b = _tuple_accss(_tmp, 2); ...

Reply:
OK then I will pause this PR, work on a new PR to translate a full generic `program` into `IL` (maybe with some TODOs), then get back to this PR. May take some days though.

(Note that AST-to-IL already handles tuples in the LHS of assigments.)